### PR TITLE
PXD-443 Feat/file submit

### PR DESCRIPTION
**Pull request by macintoshpie**

### CODE REVIEW

File: steps/sheepdogAPI.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
+'use strict';
+
+let username = process.env.INDEX_USERNAME;
```
Comment (thanh-nguyen-dang) [inline steps/sheepdogAPI.js]: can you make the code getting username, password, and creating token in the headers as a util function? Don't repeat the code in different places

File: suites/apis/submitFileTest.js [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+chai.use(chaiAsPromised);
+let expect = chai.expect;
+
+let assert = require('assert');
+
+
+Feature('SubmitFileTest');
+
+// Nodes in sorted (hierarchical) order
+const nodes = [
```
Comment (thanh-nguyen-dang) [inline suites/apis/submitFileTest.js]: it better to give every node a `key`

File: codecept.json [hunk 3]
```diff
@@ -2,7 +2,7 @@
   "output": "./output",
   "helpers": {
     "WebDriverIO": {
-      "url": "https://thanhnd.planx-pla.net",
+      "url": "https://tsummer.planx-pla.net",
```
Comment (thanh-nguyen-dang) [inline codecept.json]: could you remove `codecept.json` from github

File: suites/apis/submitFileTest.js [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    "type": "submitted_unaligned_reads",
+    "submitter_id": "submitted_unaligned_reads_001",
+    "read_groups": {
+      "submitter_id": "read_group_001"
+    }
+  }
+};
+
+Scenario('test submit file', async(I) => {
+  I.submitFile("/api/v0/submission/dev/test/", files.test_file)
```
Comment (thanh-nguyen-dang) [inline suites/apis/submitFileTest.js]: make program and project as parameters

**Review by thanh-nguyen-dang — CHANGES_REQUESTED** [review]

File: steps/utilAPI.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+'use strict';
+
+module.exports.getAccessTokenHeader = function() {
```
Comment (thanh-nguyen-dang) [inline steps/utilAPI.js]: where did you use username/password?

File: steps/fenceAPI.js [hunk 6]
```diff
@@ -2,12 +2,11 @@
 
 let username = process.env.INDEX_USERNAME;
 let password = process.env.INDEX_PASSWORD;
-let accessTokenHeaders = {
-  'Accept': 'application/json',
-  'Authorization': `bearer ${process.env.ACCESS_TOKEN}`
-};
 let indexAuth = Buffer.from(`${username}:${password}`).toString('base64');
```
Comment (thanh-nguyen-dang) [inline steps/fenceAPI.js]: check how did username/password are used here. Remove repeated code, update the util function

**Review by thanh-nguyen-dang — CHANGES_REQUESTED** [review]

File: steps/utilAPI.js (hunk 5 above)
Comment (macintoshpie) [reply steps/utilAPI.js]: From your examples it looks like username/password isn't needed in the header.
You only used the username and passwords in fenceAPI.js: addFileIndices and deleteFileIndices. For seeFileContentEqual you just used the info that I added to the util function.

File: suites/apis/submitFileTest.js [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+
+        assert(res.read_groups[0].submitter_id === files.test_file.data.read_groups.submitter_id);
+
+        I.deleteFile("/api/v0/submission/", files.test_file);
+      }
+    );
+});
+
+BeforeSuite((I) => {
+  I.addNodes("/api/v0/submission/dev/test/", Object.values(nodes))
```
Comment (thanh-nguyen-dang) [inline suites/apis/submitFileTest.js]: can it be parameterized?

File: suites/apis/submitFileTest.js [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+});
+
+AfterSuite((I) => {
+  I.deleteNodes("/api/v0/submission//dev/test/entities/", Object.values(nodes))
```
Comment (thanh-nguyen-dang) [inline suites/apis/submitFileTest.js]: can it be parameterized?

File: suites/apis/submitFileTest.js [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+      "type": "read_group",
+      "submitter_id": "read_group_001",
+      "aliquots": {
+        "submitter_id": "aliquot_001"
+      }
+    }
+  },
+];
+
+const files = {
```
Comment (thanh-nguyen-dang) [inline suites/apis/submitFileTest.js]: where is the link? could you change the filename to the one in s3 bucket?

File: suites/apis/submitFileTest.js [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+chai.use(chaiAsPromised);
+let expect = chai.expect;
+
+let assert = require('assert');
+
+
+Feature('SubmitFileTest');
+
+// Nodes, sorted hierarchically by key
+const nodes = [
```
Comment (thanh-nguyen-dang) [inline suites/apis/submitFileTest.js]: move nodes to a separated test data file because it is too large to put in a code files.

**Review by thanh-nguyen-dang — CHANGES_REQUESTED** [review]

File: steps/sheepdogAPI.js [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    `${sheep_endpoint}${file.program}/${file.project}/`, JSON.stringify(file.data), accessTokenHeader).then(
+    (res) => {
+      if (res.body.hasOwnProperty('entities')) {
+        console.log("PUT Result", res.body);
+        console.log("PUT Errors:", res.body.entities[0].errors);
+        let new_id = res.body.entities[0].id;
+        file.id = new_id;
+        let get_file = `${sheep_endpoint}${file.program}/${file.project}/export?ids=${new_id}&format=json`;
+        // get file from sheepdog to get did
+        return this.sendGetRequest(get_file, accessTokenHeader).then(
```
Comment (thanh-nguyen-dang) [inline steps/sheepdogAPI.js]: this code should be clearly split into two functions so that easier to read

**Review by thanh-nguyen-dang — CHANGES_REQUESTED** [review]

File: sample_data/submitSamples.json [hunk 12]
```diff
@@ -0,0 +1,140 @@
+{
+  "study": {
+    "key": 1,
+    "program": "dev",
```
Comment (thanh-nguyen-dang) [inline sample_data/submitSamples.json]: we still hard code program/project name here?

**Review by thanh-nguyen-dang — DISMISSED** [review]

File: sample_data/submitSamples.json (hunk 12 above)
Comment (macintoshpie) [reply sample_data/submitSamples.json]: No, that is updated in the other pull request, branch feat/graphqlQuery

**Pull request closed without merging**